DisputableVoting: Expose vote settings and results

facuspagnuolo opened the pull request:
This PR adds the following concepts to the disputable voting app:
- Allow accessing ERC20's related info: name, symbol, and decimals
- Knowing whether a vote is passing or not
- New status labels for votes including "Rejected" and "Accepted"
- New vote attributes: challenger, challenge end date, challenge ID, dispute ID, and execution date
- Allow querying the setting associated with a vote
- Formatted values for all vote results: yeas, nays, yeasPct, naysPct
- Formatted values for all pct settings: support and acceptance

It would be nice to think about how we want to start structuring the logic related to helpers/utils, specially around times, numbers, etc. Maybe we can do sth with our existing [contract-helpers](https://github.com/aragon/contract-helpers/) (cc @bpierre @0xGabi)

0xGabi commented on packages/connect-voting-disputable/src/helpers/numbers.ts:
```diff
@@ -0,0 +1,27 @@
+
+export const bn = (x: string): BigNumber => BigNumber.from(x)
+
+export const formatBn = (number: string | BigNumber, numberDecimals: string | number, formattedDecimals = 2): string => {
```
Wonder if we may wan to have `PCT_DECIMALS` as default value for `numberDecimals`.

facuspagnuolo replied:
Answered bellow in https://github.com/aragon/connect/pull/196#discussion_r472235428

0xGabi commented on packages/connect-voting-disputable/src/helpers/numbers.ts:
```diff
@@ -0,0 +1,27 @@
+  const decimals = formattedNumber.substring(decimalPosition + 1)
+  const decimalsLength = decimals.length
+
+  if (decimalsLength <= formattedDecimals) {
+    return `${formattedNumber}${'0'.repeat(formattedDecimals - decimalsLength)}`
+  }
```
I'm not super sure about this logic. I think we are missing to round the decimals when `decimalsLength <= formattedDecimals`.

facuspagnuolo replied:
Note that scenario is contemplated in the tests, and actually, if the number of decimals is lower than or equal to the number of requested decimals to format, then we don't need to round it, don't you agree?

0xGabi commented on packages/connect-voting-disputable/src/models/Setting.ts:
```diff
@@ -28,4 +26,12 @@ export default class Setting {
   }
+
+  get formattedSupportRequiredPct(): string {
+    return formatBn(this.supportRequiredPct, PCT_DECIMALS)
```
If we decided to use `PCT_DECIMALS` as default we can remove it in a few places. Or do you think is better to be explicit about it?

facuspagnuolo replied:
Well, actually note that we are using different decimals for these operations, 16 for percentages and 18 for token amounts, then I'd keep it

0xGabi commented on packages/connect-voting-disputable/src/models/Vote.ts:
```diff
@@ -50,12 +62,18 @@ export default class Vote {
+
+  get hasEnded(): boolean {
+    const now = new Date().getTime()
+    return now >= toMilliseconds(this.endDate)
```
Nice!

0xGabi commented on packages/connect-voting-disputable/src/models/Vote.ts:
```diff
@@ -50,12 +62,18 @@ export default class Vote {
-    const originalEndDate = BigNumber.from(this.startDate).add(BigNumber.from(this.duration))
-    const endDateAfterPause = originalEndDate.add(BigNumber.from(this.pauseDuration))
-    return endDateAfterPause.add(BigNumber.from(this.quietEndingExtendedSeconds)).toString()
+    const originalEndDate = bn(this.startDate).add(bn(this.duration))
+    const endDateAfterPause = originalEndDate.add(bn(this.pauseDuration))
+    return endDateAfterPause.add(bn(this.quietEndingExtendedSeconds)).toString()
```
Wonder if we also want to convert the values with `toMilliseconds` so the consumer should always expect milliseconds from the connectors.

facuspagnuolo replied:
Yes, I think that's a great idea. Although we could change all the models expose in connect at once to make sure there is no hybrid behavior, what do you think?

0xGabi replied:
Good point, let's do that!

0xGabi commented on packages/connect-voting-disputable/src/models/Vote.ts:
```diff
@@ -67,8 +85,45 @@ export default class Vote {
+  }
+
+  async formattedYeas(): Promise<string> {
+    return formatBn(this.yeas, await this._tokenDecimals())
```
What you think of having the `tokenDecimals` be part of the Vote entity at the subgraph level. I wonder if we may prevent a lot of request to fetch the decimals value that's something we are sure is not going to change after the creation of the token.

facuspagnuolo replied:
I like the idea :)

0xGabi commented on packages/connect-voting-disputable/src/models/Vote.ts:
```diff
@@ -67,8 +85,45 @@ export default class Vote {
+    return formatBn(this.yeas, await this._tokenDecimals())
+  }
+
+  async formattedYeasPct(): Promise<string> {
```
```suggestion
formattedYeasPct(): string {

0xGabi commented on packages/connect-voting-disputable/src/models/Vote.ts:
```diff
@@ -67,8 +85,45 @@ export default class Vote {
+    return formatBn(this.nays, await this._tokenDecimals())
+  }
+
+  async formattedNaysPct(): Promise<string> {
```
```suggestion
formattedNaysPct():string {

0xGabi commented on packages/connect-voting-disputable/src/models/Vote.ts:
```diff
@@ -98,4 +153,25 @@ export default class Vote {
+    return erc20.decimals
+  }
+
+  _isValuePct(value: BigNumber, total: BigNumber, pct: BigNumber): boolean {
```
What you think if change the naming of this function? I got a bit confuse in relation to what it does.

facuspagnuolo commented on packages/connect-voting-disputable/src/helpers/index.ts:
```diff
@@ -0,0 +1,2 @@
+export { toMilliseconds } from './time'
```
Yup, I already mentioned this to @sohkai... ideally I think we can start exporting some of the helpers with typescript support. However, we could probably start doing that internally for these connect packages and then export it to the contract-helpers repo right?

sohkai replied:
I'd be against re-using the helpers we have in `contract-helpers-test`, to keep that package singularly focused on making it easier for us to write tests.

We can definitely think about a `connect-helpers` package for some of these, especially as we standardize on what is a good, well-formed app connector API—I would love for us to be in a place where app connectors are implementing their dates, numbers, and token amounts in a singular, well-recommended fashion that's obvious for developers to use.

facuspagnuolo commented:
@0xGabi thx for the review! I addressed some of your comments and answered the rest of them. 
OTOH, I'm having an issue after merging the latest changes from master, maybe you are aware of what that could be about

sohkai commented on packages/connect-voting-disputable/src/models/Vote.ts:
```diff
@@ -50,25 +63,55 @@ export default class Vote {
+
+  get hasEnded(): boolean {
+    const now = new Date().getTime()
+    return now >= toMilliseconds(this.endDate)
```
Note that you can also use `Date.now()` :)

sohkai commented on packages/connect-voting-disputable/src/helpers/numbers.ts:
```diff
@@ -0,0 +1,27 @@
+import { BigNumber, utils } from 'ethers'
```
I wonder if it would be worth exposing `TokenAmount`s from app connectors' API (cc @bpierre).

(Also see next comment)

sohkai commented on packages/connect-voting-disputable/src/__test__/herlpers/numbers.test.ts:
```diff
@@ -0,0 +1,49 @@
+import { bn, formatBn } from '../../helpers'
```
Nitpick: the directory (`herlpers`—both kind of funny but scary when I initially read it as `herpies`) is typoed.

facuspagnuolo commented:
@0xGabi added some more stuff! let me know your thoughts!

0xGabi commented on packages/connect-voting-disputable/subgraph/src/DisputableVoting.ts:
```diff
@@ -145,6 +146,16 @@ export function handleChangeRepresentative(event: ChangeRepresentativeEvent): vo
   voter.save()
 }
 
+function isAccepted(yeas: BigInt, nays: BigInt, votingPower: BigInt, settingId: string, pctBase: BigInt): boolean {
```
Look much better to manipulate the `BigNumber`s at the subgraph level 👍
